### DISCUSSION TITLE
Issue in fitting model and finding optimal learning rate parameter

### DISCUSSION THREAD

**Discussion by geeksouvik** [Lightning-AI/pytorch-lightning#10341] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
```py
following is the error: NotImplementedError: `val_dataloader` must be implemented to be used with the Lightning Trainer

LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-11-263e8be26564> in <module>()
      3     tft,
      4     train_dataloader=train_dataloader,
----> 5     val_dataloaders=val_dataloader,
      6 )

11 frames
/usr/local/lib/python3.7/dist-packages/pytorch_lightning/core/hooks.py in val_dataloader(self)
    590             will have an argument ``dataloader_idx`` which matches the order here.
    591         """
--> 592         raise NotImplementedError("`val_dataloader` must be implemented to be used with the Lightning Trainer")
    593 
    594     def predict_dataloader(self) -> EVAL_DATALOADERS:

NotImplementedError: `val_dataloader` must be implemented to be used with the Lightning Trainer
```

```py
trainer.fit(
    tft,
    train_dataloader=train_dataloader,
    val_dataloaders=val_dataloader,
)
```

**Comment by furbyhaxx** [comment, marked as answer]
I think it's a problem with pytorch-lightning==1.5.0.
Had this problem too with code that worked before recreating my venv and the difference was version 1.5.0 release on 2. Nov.
Switched to 1.4.9 and it worked again.
Checked 1.5.0rc1 and it did not work either.

**Reply by paapu88** [reply]
Thanks!!! So typical with pytorch-lightning. I dislike it and its video documentations but its best what you can get...

**Reply by geeksouvik** [reply]
Yea.. right it's the problem with the 1.5.0 version. downgraded the package and it's working.

**Comment by ilkerkesen** [comment]
I realized that you cannot use `self.x_dataloader()` methods inside the `LightningModule` implementation anymore. I need to call `self.val_dataloader()` method to post-process the model predictions. I guess this has been changed with the new version. Right now, you can access the datamodule object with `datamodule` field within the lightning module implementation (e.g. `self.datamodule`, `self.datamodule.val_dataloader()` or `self.datamodule.val_data`).

**Reply by gsoykan** [reply]
as of version 1.7.7, here is what I'm doing similar to your solution => 
`self.trainer.datamodule.train_dataloader()`

**Comment by ananthsub** [comment]
@geeksouvik @paapu88 @ilkerkesen : 

the `self.x_dataloader()` are optional methods defined in the LightningModule, in case your module and data are not separable. The default implementation of these methods was changed from logging a warning message that these funtions should be implemented to raising a `NotImplementedError` to be explicit in case you calling something which you haven't defined. 

Though the naming is similar, this is very different from accessing an already instantiated dataloader which you've passed to the trainer outside of the LightningModule's definition. If your model depends on the data, I'd recommend making it an input to your LightningModule constructor, and then pass that to the Trainer for `fit` and `test`

**Reply by sherandoresearch** [reply]
I am new to lightning and trying to get up to speed. I am using it with pytorch-forecasting and attempting to run this example: https://pytorch-forecasting.readthedocs.io/en/latest/tutorials/stallion.html

At step 8 I get the Not Implemented Error for the val_dataloader. Sorry to ask for a more explicit explanation, but where are you suggesting I include the loaders? at pl.Trainer?
